### PR TITLE
[Merged by Bors] - feat: response trace streaming (PL-909)

### PR DESCRIPTION
**Pull request by trs**
Introduces a new route to interact with your assistant, enabling server-sent events to stream traces back as they are generated.
This also hooks up with LLMs to stream back partial traces to enable the "typing" like responses of LLMs.

Depends on: https://github.com/voiceflow/libs/pull/520

### CODE REVIEW

File: lib/controllers/interact/index.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+class InteractController extends AbstractController {
+  async stream(
+    req: Request<InteractRequestParams, InteractRequestBody, InteractRequestHeaders, InteractRequestQuery>,
+    res: Response
+  ) {
+    const params = await InteractRequestParams.parseAsync(req.params);
+    const body = await InteractRequestBody.parseAsync(req.body);
+    // const _headers = await InteractRequestHeaders.parseAsync(req.headers);
+    // const _query = await InteractRequestQuery.parseAsync(req.query);
```
Comment (theProf) [inline lib/controllers/interact/index.ts:34]: commented code

File: lib/controllers/interact/index.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
+/**
+ * [[include:contextHandlers.md]]
+ * @packageDocumentation
+ */
+
+import { BaseRequest, RuntimeLogs } from '@voiceflow/base-types';
+import { createSession } from 'better-sse';
+import { match } from 'ts-pattern';
+
+// import { isEmpty, merge } from "lodash";
```
Comment (theProf) [inline lib/controllers/interact/index.ts:10]: commented code

File: runtime/lib/Runtime/cycleHandler.ts [hunk 3]
```diff
@@ -1,15 +1,21 @@
-/* eslint-disable no-await-in-loop */
+/* eslint-disable no-await-in-loop, sonarjs/cognitive-complexity */
```
[the remark this answers is not in the document]
Comment (theProf) [reply runtime/lib/Runtime/cycleHandler.ts:1]: i know this isn't your doing, it just....quite the loop logic in there...

File: lib/services/nlu/utils.ts [hunk 4]
```diff
@@ -145,7 +145,7 @@ export const getAvailableIntentsAndEntities = async (
   availableEntities: Set<string>;
   bypass?: boolean;
 }> => {
-  const runtime = runtimeManager.createClient(context.data.api).createRuntime({
+  const runtime = runtimeManager.createClient(context.data.api, () => undefined).createRuntime({
```
[the remark this answers is not in the document]
Comment (trs) [reply lib/services/nlu/utils.ts:148]: I had this at first but our outdated eslint rules didn't like us using `void`, so this was just the path of least resistance lol

File: lib/services/merge-completion/guards.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import { BaseTrace } from '@voiceflow/base-types';
+
+export const isCompletionStartTrace = (trace: BaseTrace.AnyTrace): trace is BaseTrace.CompletionStartTrace =>
+  trace.type === BaseTrace.TraceType.COMPLETION_START;
+
+export const isCompletionContinueTrace = (trace: BaseTrace.AnyTrace): trace is BaseTrace.CompletionContinueTrace =>
+  trace.type === BaseTrace.TraceType.COMPLETION_CONTINUE;
+
+export const isCompletionEndTrace = (trace: BaseTrace.AnyTrace): trace is BaseTrace.CompletionEndTrace =>
+  trace.type === BaseTrace.TraceType.COMPLETION_END;
```
[the remark this answers is not in the document]
Comment (trs) [reply lib/services/merge-completion/guards.ts:10]: Zod is the way at least medium term

File: lib/services/index.ts [hunk 6]
```diff
@@ -53,7 +55,13 @@ const buildServices = (config: Config, clients: ClientMap): FullServiceMap => {
   } as FullServiceMap;
 
   // order here matters, services defined after runtime are not available inside of it
+  if (config.SESSIONS_SOURCE === Source.LOCAL) {
+    services.session = new LocalSession(services, config);
+  } else if (config.SESSIONS_SOURCE === Source.MONGO) {
+    services.session = new MongoSession(services, config);
+  }
```
[the remark this answers is not in the document]
Comment (trs) [reply lib/services/index.ts:62]: I can move it later, really only the interact service needs it

File: runtime/lib/Runtime/cycleHandler.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (trs) [reply runtime/lib/Runtime/cycleHandler.ts:1]: 100%

File: lib/services/merge-completion/index.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+      return acc;
+    }, []);
+
+    return {
+      ...context,
+      trace,
+    };
+  }
+}
```
Comment (theProf) [inline lib/services/merge-completion/index.ts:61]:
```
export const isCompletionTrace = (trace: BaseTrace.AnyTrace) =>
  isCompletionStartTrace(trace) || isCompletionContinueTrace(trace) || isCompletionEndTrace(trace);

const processCompletion =
  ({ project, version, state: { variables } }: Context) =>
    (acc: BaseTrace.AnyTrace, trace: BaseTrace.AnyTrace, index: number) => {
      switch (true) {
        case isCompletionStartTrace(trace): {
          if (index !== 0) {
            throw new VError('Start Trace out of order');
          }
          return trace;
        }
        case isCompletionContinueTrace(trace): {
          if (!isCompletionStartTrace(acc)) {
            throw new VError('Continue Trace out of order');
          }
          acc.payload.completion += trace.payload.completion;
          return acc;
        }
        case isCompletionEndTrace(trace): {
          if (!isCompletionStartTrace(acc)) {
            throw new VError('End Trace out of order');
          }
          const output = generateOutput(acc.payload.completion, project, acc.payload.voice);
          return getOutputTrace({
            output,
            ai: true,
            variables: new Store(variables),
            version,
            delay: acc.payload.delay,
          });
        }
        default: {
          throw new VError('Non Completion Trace in completion processing pipeline');
        }
      }
    };

@injectServices({ utils })
export default class MergeCompletion extends AbstractManager<{ utils: typeof utils }> implements ContextHandler {
  async handle(context: Context) {
    if (!context.trace) return context;

    const trace$ = from(context.trace).pipe(
      // outputs 2 observable streams:
      // 1. observable stream of non-completion traces
      // 2. observable stream of _only_ completion traces
      groupBy(isCompletionTrace),
      mergeMap((group$) =>
        !group$.key
          ? // stream 1 needs no additional processing, just emit
          group$
          : // stream 2 reduces 1 start->end completion trace at a time, emitting a single observable per re-put-together trace
          group$.pipe(
            scan(processCompletion(context)),
            filter((x) => x.type === 'end')
          )
      ),
      toArray()
    );

    return {
      ...context,
      trace: await lastValueFrom(trace$),
    };
  }
}```

File: backend/api/routers/v2/interact.ts [hunk 8]
```diff
@@ -10,7 +10,12 @@ export default (middlewares: MiddlewareMap, controllers: ControllerMap) => {
   router.use(bodyParser.json({ limit: BODY_PARSER_SIZE_LIMIT }));
   router.use(middlewares.rateLimit.verify);
 
-  router.post('/:projectID/:tag/stream', middlewares.project.getProjectEnvironment, controllers.interact.stream as any);
+  router.post(
+    '/:projectID/:tag/stream',
+    middlewares.auth.authorize(['project:READ', 'project.versions:READ']),
```
Comment (theProf) [inline backend/api/routers/v2/interact.ts:15]: nice catch

**Review by theProf — APPROVED** [review]

**Comment by trs** [comment]
bors merge

File: backend/api/routers/v2/interact.ts (hunk 8 above)
Comment (trs) [reply backend/api/routers/v2/interact.ts:15]: 🫣

**Comment by trs** [comment]
bors retry

**Pull request closed without merging**